projects/core/src/site-context/config/config-loader/site-context-config-initializer.ts

```diff
@@ -1,44 +1,131 @@
+import { Injectable } from '@angular/core';
+import { ConfigInitializer } from 'projects/core/src/config';
+import { Observable } from 'rxjs';
+import { filter, map, take } from 'rxjs/operators';
+import { BaseSite } from '../../../model/misc.model';
+import { JavaRegExpConverter } from '../../../util/java-reg-exp-converter/java-reg-exp-converter';
+import { WindowRef } from '../../../window/window-ref';
+import { BaseSiteService } from '../../facade/base-site.service';
 import {
-  ConfigInitializer,
-  CONFIG_INITIALIZER,
-} from '../../../config/config-initializer/config-initializer';
-import { FeatureConfigService } from '../../../features-config/services/feature-config.service';
-import { SiteContextConfigLoaderService } from './../../config/config-loader/site-context-config-loader.service';
-import { SiteContextConfig } from './../../config/site-context-config';
-import { BASE_SITE_CONTEXT_ID } from './../../providers/context-ids';
-
-/**
- * Initializes the site context config
- */
-export function initSiteContextConfig(
-  configLoader: SiteContextConfigLoaderService,
-  config: SiteContextConfig,
-  featureConfigService: FeatureConfigService
-): ConfigInitializer | null {
-  // TODO(#11515): in 4.0 drop the feature level guard
-  if (!featureConfigService.isLevel('3.2')) {
-    return null;
+  BASE_SITE_CONTEXT_ID,
+  CURRENCY_CONTEXT_ID,
+  LANGUAGE_CONTEXT_ID,
+  THEME_CONTEXT_ID,
+} from '../../providers/context-ids';
+import { SiteContextConfig } from '../site-context-config';
+
+@Injectable({ providedIn: 'root' })
+export class SiteContextConfigInitializer implements ConfigInitializer {
+  readonly scopes = ['context'];
+  readonly configFactory = () => this.resolveConfig().toPromise();
+
+  constructor(
+    protected baseSiteService: BaseSiteService,
+    protected javaRegExpConverter: JavaRegExpConverter,
+    protected winRef: WindowRef
+  ) {}
+
+  private get currentUrl(): string {
+    return this.winRef.location.href as string;
   }
 
   /**
-   * Load config for `context` from backend only when there is no static config for `context.baseSite`
+   * Emits the site context config basing on the current base site data.
+   *
+   * Completes after emitting the value.
    */
-  if (!config.context || !config.context[BASE_SITE_CONTEXT_ID]) {
-    return {
-      scopes: ['context'],
-      configFactory: () => configLoader.loadConfig().toPromise(),
-    };
+  protected resolveConfig(): Observable<SiteContextConfig> {
```

Comment:
As you noted offline, now we're lacking the logic for feature level (and also check if the base site was not already set). We might add in this method:
```ts
    // TODO(#11515): remove it in 4.0
    if(!this.featureConfigService.isLevel('3.2')){
      return of({});
    } 

    // Load config for `context` from backend only when there is no static config for `context.baseSite`
    if(this.config?.context?.[BASE_SITE_CONTEXT_ID]){
      return of({});
    }
 ```